packages/dds/merge-tree/src/test/revertibleFarm.spec.ts

```diff
@@ -0,0 +1,182 @@
+                                    minLen,
+                                    defaultOptions.operations));
+                            }
+
+                            if (opsWithRevert > 0) {
+                                // add modifications from another client
+                                msgs.push(...generateOperationMessagesForClients(
+                                    mt,
+                                    seq,
+                                    [clients.A, clients.C],
+                                    logger,
+                                    opsWithRevert,
+                                    minLen,
+                                    defaultOptions.operations));
+                            }
+
+                            let redoBaseText: string | undefined;
+                            if (ackBeforeRevert !== "None") {
+                                const ackAll = ackBeforeRevert === "All";
+                                seq = applyMessages(
+                                    seq,
+                                    msgs.splice(
+                                        0,
+                                        ackAll
+                                            ? msgs.length
+                                            : random.integer(0, Math.floor(msgs.length / 2))(mt)),
+                                    clients.all,
+                                    logger);
+                                if (ackAll) {
+                                    redoBaseText = logger.validate({ errorPrefix: "Before Revert Ack" });
+                                }
+                            }
+
+                            try {
+                                revertMergeTreeDeltaRevertibles(clientBDriver, clientB_Revertibles.splice(0));
+                                seq = applyMessages(seq, msgs.splice(0), clients.all, logger);
+                            } catch (e) {
+                                throw logger.addLogsToError(e);
+                            }
+                            logger.validate({
+                                clear: true,
+                                baseText: opsWithRevert === 0 ? undoBaseText : undefined,
+                                errorPrefix: "After Revert (undo)",
+                            });
+
+                            try {
+                                // reset the callback before the final revert
+                                // to avoid accruing any new detached references
+                                clients.B.mergeTreeDeltaCallback = oldCallback;
+                                revertMergeTreeDeltaRevertibles(clientBDriver, clientB_Revertibles.splice(0));
+                                seq = applyMessages(seq, msgs.splice(0), clients.all, logger);
+
+                                // validate that there are no lingering/leaked references in detached references
+                                const detachedReferences =
```

Comment:
in addition to these, maybe also worth validating that there aren't any references remaining on segments in the merge tree which have type "mergeTreeDeltaRevertible". It would *probably* cause the tests to fail already, but if there was some sort of problem that happened only with concurrent ops (so that we don't validate the text content as well) i'm not so sure.

Reply:
good call on this. found a small memory leak related to empty tracking groups!